aptos-move/block-executor/src/view.rs

```diff
@@ -811,14 +1159,123 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
+                Some(ValueWithLayout::RawFromStorage(_)) => Some(Err(code_invariant_error(
+                    "Cannot exchange value that was not exchanged before",
+                ))),
+                None => None,
+            })
+            .collect()
+    }
+
+    fn get_group_reads_needing_exchange_parallel(
+        &self,
+        parallel_state: &ParallelState<'a, T, X>,
+        delayed_write_set_keys: &HashSet<T::Identifier>,
+        skip: &HashSet<T::Key>,
+    ) -> Result<BTreeMap<T::Key, (T::Value, u64)>, PanicError> {
+        parallel_state
+            .captured_reads
+            .borrow()
+            .get_group_read_values_with_delayed_fields(skip)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .flat_map(|(key, group_read)| {
+                let GroupRead { inner_reads, .. } = group_read;
+                let mut resources_needing_delayed_field_exchange = false;
+                for data_read in inner_reads.values() {
+                    if let DataRead::Versioned(_version, value, Some(layout)) = data_read {
+                        if let Some(bytes) = value.bytes() {
+                            let identifiers_in_read = self
+                                .extract_identifiers_from_value(bytes, layout.as_ref())
+                                .unwrap();
+                            if !delayed_write_set_keys.is_disjoint(&identifiers_in_read) {
+                                // TODO[agg_v2](optimize): Is it possible to avoid clones here?
+                                resources_needing_delayed_field_exchange = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+                if !resources_needing_delayed_field_exchange {
+                    return None;
+                }
+
+                // TODO[agg_v2](fix): Is it guaranteed that metadata is not None?
+                if let Ok(Some(maybe_metadata)) = self.get_resource_state_value_metadata(&key) {
+                    let maybe_metadata = maybe_metadata.map(
+                        |metadata: aptos_types::state_store::state_value::StateValueMetadata| {
+                            StateValue::new_with_metadata(Bytes::new(), metadata)
+                        },
                     );
+                    if let Ok(GroupReadResult::Size(group_size)) =
+                        parallel_state.read_group_size(&key, self.txn_idx)
+                    {
+                        let metadata_op: T::Value =
+                            TransactionWrite::from_state_value(maybe_metadata);
+                        if let Some(metadata_op) = metadata_op.convert_read_to_modification() {
+                            return Some(Ok((key.clone(), (metadata_op, group_size))));
+                        }
+                    }
+                }
+                // TODO[agg_v2](fix): Is this a code invariant error?
```

Comment:
code invariant error makes sense here I guess, but doesn't distinguish between the two errors: when somehow we didn't get the size, and when we somehow didn't get metadata.

Reply:
Separated the code invariant errors.